docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,377 @@
+results regardless of the configured execution mode.  By enabling `BATCH`
+execution, we allow Flink to apply additional optimizations that we can only do
+work when we know that our input is bounded. For example, different
+join/aggregation strategies can be used, in addition to a different shuffle
+implementation that allows more efficient failure recovery behavior. We will go
+into some of the details of the execution behavior below.
+
+* This will be replaced by the TOC
+{:toc}
+
+## When can/should I use BATCH execution mode?
+
+The `BATCH` execution mode can only be used for Jobs/Flink Programs that are
+_bounded_. Boundedness is a property of a data source that tells us whether all
+the input coming from that source is known before execution or whether new data
+will show up, potentially indefinitely. A job, in turn, is bounded if all its
+sources are bounded, and unbounded otherwise.
+
+`STREAMING` execution mode, on the other hand, can be used for both bounded and
+unbounded jobs.
+
+As a rule of thumb, you should be using `BATCH` execution mode when your program
+is bounded because this will be more efficient. You have to use `STREAMING`
+execution mode when your program is unbounded because only this mode is general
+enough to be able to deal with continuous data streams.
+
+One obvious outlier case is when you want to use a bounded job to bootstrap
+some job state that you then want to use in an unbounded job. For example, by
+running a bounded job using `STREAMING` mode, taking a savepoint, and then
+restoring that savepoint on an unbounded job. This is a very specific use case
+and one that might soon become obsolete when we allow producing a savepoint as
+additional output of a `BATCH` execution job.
+
+Another case where you might run a bounded job using `STREAMING` mode is when
+writing tests for code that will eventually run with unbounded sources. For
+testing it can be more natural to use a bounded source in those cases.
+
+## Configuring BATCH execution mode
+
+The execution mode can be configured via the `execution.runtime-mode` setting.
+There are three possible values:
+
+ - `STREAMING`: The classic DataStream execution mode (default)
+ - `BATCH`: Batch-style execution on the DataStream API
+ - `AUTOMATIC`: Let the system decide based on the boundedness of the sources
+
+This can be configured via command line parameters of `bin/flink run ...`, or
+programmatically when creating/configuring the `StreamExecutionEnvironment`.
+
+Here's how you can configure the execution mode via the command line:
+
+```bash
+$ bin/flink run -Dexecution.runtime-mode=BATCH examples/streaming/WordCount.jar
+```
+
+This example shows how you can configure the execution mode in code:
+
+ ```java
+Configuration config = new Configuration();
+config.set(ExecutionOptions.RUNTIME_MODE, RuntimeExecutionMode.BATCH);
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+ ```
+
```

Comment:
I would add: 

```
or 

```java
StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(config);
env.setRuntimeMode(RuntimeExecutionMode.BATCH);
 ```

<span class="label label-danger">Important Note</span>: We recommend users to NOT use this method
but set the `execution.runtime-mode` using the command-line when submitting the application. Keeping 
the application code configuration-free allows for more flexibility as the same application will be
able to be executed in any execution mode.

```

Reply:
Thanks, that's a good suggestion!